osm2streets/src/lib.rs

```diff
@@ -200,40 +199,18 @@ impl StreetNetwork {
         self.intersections[&i].roads.clone()
     }
 
-    /// (Intersection polygon, polygons for roads, list of labeled polygons to debug)
-    #[allow(clippy::type_complexity)]
-    pub fn preview_intersection(
-        &self,
-        id: osm::NodeID,
-    ) -> Result<(Polygon, Vec<Polygon>, Vec<(String, Polygon)>)> {
-        let mut input_roads = Vec::new();
-        for r in self.roads_per_intersection(id) {
-            input_roads.push(initial::Road::new(self, r).to_input_road());
-        }
-        let results = intersection_polygon(
-            id,
-            input_roads,
-            // This'll be empty unless we've called merge_short_road
-            &self.intersections[&id].trim_roads_for_merging,
-        )?;
-        Ok((
-            results.intersection_polygon,
-            results
-                .trimmed_center_pts
-                .into_values()
-                .map(|(pl, half_width)| pl.make_polygons(2.0 * half_width))
-                .collect(),
-            results.debug,
-        ))
-    }
-
-    /// Generate the trimmed `PolyLine` for a single Road by calculating both intersections
-    pub fn trimmed_road_geometry(&self, road_id: OriginalRoad) -> Result<PolyLine> {
+    /// This calculates a road's `trimmed_center_line` early, before
+    /// `Transformation::GenerateIntersectionGeometry` has run. Use sparingly.
+    pub(crate) fn estimate_trimmed_geometry(&self, road_id: OriginalRoad) -> Result<PolyLine> {
         // First trim at one of the endpoints
         let trimmed_center_pts = {
             let mut input_roads = Vec::new();
             for r in self.roads_per_intersection(road_id.i1) {
-                input_roads.push(initial::Road::new(self, r).to_input_road());
+                let road = &self.roads[&r];
+                // trimmed_center_line hasn't been initialized yet, so override this
```

Comment:
I thought about setting `trimmed_center_line = untrimmed_center_line` in `Road::new`, but all the places that modify `untrimmed_center_line` would need to keep things in sync... This is the only place where we need to worry about this

Reply:
Alternative to the cached getter idea: Keep the data stored within a single `Road` or `Intersection` internally consistent and safe to use, but allow the resulting network to have inconsistent (overlapping etc.) geometry until the appropriate transformations have run.

I imaging setting `trimmed_center_line = untrimmed_center_line` and `{start,end)_trim_distance = 0` in `Road::new` like you suggest, and `polygon = <some little square/circle at the center>` in `Intersection::new`. That way it is always safe to use those fields in code, and the ordering of transformation only matters for interpreting the values.

I imagine instantiating all of the `Roads` and `Intersections` from the osm data and being able to visualise it immediately (using the normal debug vis mechanisms). First a transformation accounts for placement. Then a transformation looks at all of the intersections and trims roads back just enough to generate sensible intersection geom. Then a transformation collapses roads with a negative trimmed distance (therefore merging neighbouring intersections that have overlapped). Then movements can be calculated and intersections classified. Then all manor of heuristic transformations can take place, like parallel way snapping, like collapsing short `Road`s between two `Cross`ing `Intersection`s (because `Cross` implies roads markings are interrupted, so if the road is short it probably doesn't have it's own road markings, so it's probably internal to the intersections), or like increasing the trim distance based on the kind of intersection or the position of pedestrian crossings/stop lines.

I haven't got a good picture of how the existing transformations fit together yet, I will wrap my head around it as I work on placement tags and on trying to tweak intersections based on `complexity`/`conflic_level`.

Reply:
I like this idea. It'd also mean the system is totally self-explaining through the debug steps -- huzzah for slide decks for free! Explain how things work by stepping through.

> Then a transformation looks at all of the intersections and trims roads back just enough to generate sensible intersection geom.

I've been picturing the geometry calculation as always happening at the very end based on how the other transformations work currently: (almost) none of them use trimmed distances. But that's just a historic assumption.

> Then a transformation collapses roads with a negative trimmed distance (therefore merging neighbouring intersections that have overlapped).

Hmm, nothing produces negative trim distances today, but that's probably an assumption baked into the intersection geometry algo. This sounds like a nicer way to detect the case of two intersections overlapping than actually checking the polygons themselves.

> like collapsing short Roads between two Crossing Intersections (because Cross implies roads markings are interrupted, so if the road is short it probably doesn't have it's own road markings, so it's probably internal to the intersections)

!! This is the great fresh new ideas I'm super hyped for. If you dig around, you'll probably find 2 or 3 incorrect attempts to auto-detect short roads for merging. None of those attempts have had anything resembling intersection complexity as an input.

> I haven't got a good picture of how the existing transformations fit together yet

I will try to start some more substantial docs about how everything works currently. Please do ask questions in the meantime.

Lots of new ideas / actions in this comment, will keep starred in my own notes for now